packages/nimble-components/src/table/models/keyboard-navigation-manager.ts

```diff
@@ -0,0 +1,1243 @@
+import { Notifier, Subscriber, Observable } from '@microsoft/fast-element';
+import {
+    keyArrowDown,
+    keyArrowLeft,
+    keyArrowRight,
+    keyArrowUp,
+    keyEnd,
+    keyEnter,
+    keyEscape,
+    keyFunction2,
+    keyHome,
+    keyPageDown,
+    keyPageUp,
+    keySpace,
+    keyTab
+} from '@microsoft/fast-web-utilities';
+import { FoundationElement } from '@microsoft/fast-foundation';
+import type { ScrollToOptions } from '@tanstack/virtual-core';
+import type { Table } from '..';
+import {
+    TableActionMenuToggleEventDetail,
+    TableFocusType,
+    TableHeaderFocusableElements,
+    TableRowFocusableElements,
+    type TableRecord
+} from '../types';
+import type { Virtualizer } from './virtualizer';
+import { TableGroupRow } from '../components/group-row';
+import { TableRow } from '../components/row';
+import { TableCell } from '../components/cell';
+import { MenuButton } from '../../menu-button';
+import { tableHeaderTag } from '../components/header';
+import { TableCellView } from '../../table-column/base/cell-view';
+
+interface TableFocusState {
+    focusType: TableFocusType;
+    headerActionIndex?: number;
+    rowIndex?: number;
+    columnIndex?: number;
+    cellContentIndex?: number;
+}
+
+/**
+ * Manages the keyboard navigation and focus within the table.
+ * @internal
+ */
+export class KeyboardNavigationManager<TData extends TableRecord>
+implements Subscriber {
+    private focusType: TableFocusType = TableFocusType.none;
+    private headerActionIndex = -1;
+    private rowIndex = -1;
+    private cellContentIndex = -1;
+    private columnIndex = -1;
```

Comment:
Is there a potential for wonkiness if a column gets hidden while focus is within the table, particularly if the focus is on a cell in a column that is past the one that is hidden? This was difficult to make happen easily in Storybook as once you hide a column in dev tools, focus is lost.

Reply:
In an attempt to keep things simpler, my overall strategy is to reset focus back to the default location (i.e. start back at the first top/left focusable element of the table) when table configurations change (in a way that would affect focus). (That code path is `table.update()` -> `keyboardNavigationManager.resetFocusState()`)

That being said, it looks like I overlooked adding `columnHidden` to that code path, so I'll do that. (I probably also could use a test with a column with that set to make sure the navigation keys work as expected)

Reply:
Updated `columnHidden` to reset focus state too. Added tests to cover that hiding a column/ changing selection mode/ adding an action menu should reset focus.